api/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,159 @@
+	Parameters []HelmParameter `json:"parameters,omitempty" mapstructure:"parameters,omitempty"`
+}
+
+// HelmParameter is a parameter passed to helm template during manifest generation
+type HelmParameter struct {
+	// Name is the name of the Helm parameter
+	Name string `json:"name,omitempty" mapstructure:"name,omitempty"`
+	// Value is the value for the Helm parameter
+	Value string `json:"value,omitempty" mapstructure:"value,omitempty"`
+}
+
+// RawSource holds raw specific options
+type RawSource struct{}
+
+// ExplicitType returns the type (e.g., Helm, Kustomize, etc.) of the application. If either none or multiple types are defined, returns an error.
+func (gitSource *GitSource) ExplicitType() (SourceType, error) {
+	var appTypes []SourceType
+	if gitSource.Kustomize != nil {
+		appTypes = append(appTypes, SourceTypeKustomize)
+	}
+	if gitSource.Helm != nil {
+		appTypes = append(appTypes, SourceTypeHelm)
+	}
+	if gitSource.Raw != nil {
+		appTypes = append(appTypes, SourceTypeRaw)
+	}
+	if len(appTypes) == 0 {
+		// Fallback to a raw source type if a user has not specified anything.
+		return SourceTypeRaw, nil
+	}
+	if len(appTypes) > 1 {
+		typeNames := make([]string, len(appTypes))
+		for i := range appTypes {
+			typeNames[i] = string(appTypes[i])
+		}
+		return "", fmt.Errorf("multiple sources defined: %s", strings.Join(typeNames, ","))
+	}
+	appType := appTypes[0]
+	return appType, nil
+}
+
+type RepoCredential struct {
+	URL            string          `json:"url" mapstructure:"url"`
+	HTTPCredential *HTTPCredential `json:"httpCredential,omitempty" mapstructure:"httpCredential,omitempty"`
+	SSHCredential  *SSHCredential  `json:"sshCredential,omitempty" mapstructure:"sshCredential,omitempty"`
+	TLS            *TLS            `json:"tls,omitempty" mapstructure:"tls,omitempty"`
+}
+
+type HTTPCredential struct {
+	Username string            `json:"username" mapstructure:"username"`
+	Password SecretKeySelector `json:"password" mapstructure:"password"`
+}
+
+type SSHCredential struct {
+	SSHKey SecretKeySelector `json:"SSHKey" mapstructure:"SSHKey"`
+}
+
+type TLS struct {
+	InsecureSkipVerify bool `json:"insecureSkipVerify" mapstructure:"insecureSkipVerify"`
+}
+
+type SecretKeySelector struct {
+	corev1.ObjectReference `json:",inline" mapstructure:",squash"` // for viper to correctly parse the config
+	Key                    string                                  `json:"key" mapstructure:"key"`
+	Optional               *bool                                   `json:"optional,omitempty" mapstructure:"optional,omitempty"`
+}
+
+type SingleClusterGenerator struct {
+	Values map[string]string `json:"values" mapstructure:"values"`
+}
+
+// multiple files containing key/value pairs: subsequent entries can override earlier entries
+type MultiClusterFileGenerator struct {
+	Files []*MultiClusterFile `json:"files" mapstructure:"files"`
+}
+
+type MultiClusterFile struct {
```

Comment:
nit: add a comment

Reply:
I realized the struct was more generic so I renamed it to `CredentialedGitLocation` and added a comment for it there